source/server/drain_manager_impl.cc

```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
     return false;
   }
 
-  // We use the tick time as in increasing chance that we shutdown connections.
-  return static_cast<uint64_t>(drain_time_completed_.load()) >
-         (server_.random().random() % server_.options().drainTime().count());
-}
-
-void DrainManagerImpl::drainSequenceTick() {
-  ENVOY_LOG(trace, "drain tick #{}", drain_time_completed_.load());
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
+  const auto remaining_time = std::chrono::duration_cast<std::chrono::seconds>(
+      drain_deadline_ - server_.dispatcher().timeSource().monotonicTime());
```

Comment:
what if time has moved past the drain deadline?  You'll get a negative number from your subtraction. Will this code work?

Can you strobe the monotonicTime() and compare it to drain_deadline_ before subtracting?

Reply:
I don't understand what you mean by strobing, but I added tests to cover the negative case. The subtraction works out, elapsed time will exceed the drain deadline, and therefore always return true regardless of the `random() % drain_timeout`.

Reply:
strobing just means to capture the current monotonicTime() once in a temp, rather than calling it twice (once for the compare and once for the subtract), with its value potentially changing between the two calls.

I was searching for a EE reference for that term but can't find the usage I always thought of, which was analogous to a strobe in photography where you can capture a moment in time even with a fast moving object.

Reply:
Done.

\* Put the current time into a variable
\* Return true if current time is past the deadline
\* Extra testing to ensure random() is only called when elapsed time is less the drain timeout, even if the random() returned would cause drainClose() to return true